pkg/api/rest/request.go

```diff
@@ -0,0 +1,397 @@
+		r.maxRetries = maxRetries
+	} else {
+		r.maxRetries = defaultMaxRetries
+	}
+	return r
+}
+
+// Body makes http request use obj as its body.
+func (r *Request) Body(obj interface{}) *Request {
+	if r.err != nil {
+		return r
+	}
+
+	// only supports two types now:
+	//   1. io.Reader
+	//   2. type which can be json marshalled
+	if rd, ok := obj.(io.Reader); ok {
+		r.body = rd
+	} else {
+		b, err := json.Marshal(obj)
+		if err != nil {
+			r.err = err
+			return r
+		}
+		r.body = bytes.NewReader(b)
+		r.SetHeader("Content-Type", "application/json")
+	}
+	return r
+}
+
+// URL returns the current working URL.
+func (r *Request) URL() *url.URL {
+	p := r.pathPrefix
+
+	finalURL := &url.URL{}
+	if r.c.base != nil {
+		*finalURL = *r.c.base
+	}
+	finalURL.Path = p
+
+	query := url.Values{}
+	for key, values := range r.params {
+		for _, value := range values {
+			query.Add(key, value)
+		}
+	}
+
+	finalURL.RawQuery = query.Encode()
+	return finalURL
+}
+
+func (r *Request) newHttpRequest(ctx context.Context) (*http.Request, error) {
+	url := r.URL().String()
+	req, err := http.NewRequest(r.verb, url, r.body)
+	if err != nil {
+		return nil, err
+	}
+	req = req.WithContext(ctx)
+	req.Header = r.headers
+	return req, nil
+}
+
+// Do formats and executes the request.
+func (r *Request) Do(ctx context.Context) (res *Result) {
+	if r.err != nil {
+		log.Info("error in request", zap.Error(r.err))
+		return &Result{err: r.err}
+	}
+
+	client := r.c.Client
+	if client == nil {
+		client = &httputil.Client{}
+	}
+
+	if r.timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, r.timeout)
+		defer cancel()
+	}
+
+	baseDelay := r.backoffBaseDelay.Milliseconds()
+	if baseDelay == 0 {
+		baseDelay = defaultBackoffBaseDelayInMs
+	}
+	maxDelay := r.backoffMaxDelay.Milliseconds()
+	if maxDelay == 0 {
+		maxDelay = defaultBackoffMaxDelayInMs
+	}
+	maxRetries := r.maxRetries
+	if maxRetries == 0 {
+		maxRetries = defaultMaxRetries
+	}
+
+	err := retry.Do(ctx, func() error {
+		req, err := r.newHttpRequest(ctx)
+		if err != nil {
+			return err
+		}
+		// rewind the request body when r.body is not nil
+		if seeker, ok := r.body.(io.Seeker); ok && r.body != nil {
+			if _, err := seeker.Seek(0, 0); err != nil {
+				return terrors.Annotatef(err, "failed to seek to the beginning of request body:%v", r)
+			}
+		}
+
+		resp, err := client.Do(req)
+		if err != nil {
+			log.Error("failed to send a http request", zap.Error(err))
+			return err
+		}
+
+		defer func() {
+			if resp == nil {
+				return
+			}
+			// close the body to let the TCP connection be reused after reconnecting
+			io.Copy(ioutil.Discard, resp.Body)
+			resp.Body.Close()
+		}()
+
+		res = r.checkResponse(resp)
+		if res.Error() != nil {
+			return res.Error()
+		}
+		return nil
+	},
+		retry.WithBackoffBaseDelay(baseDelay),
+		retry.WithBackoffMaxDelay(maxDelay),
+		retry.WithMaxTries(maxRetries),
+		retry.WithIsRetryableErr(cerrors.IsRetryableError),
+	)
```

Comment:
Can you discuss the consequences of retrying on potentially non-idempotent API calls?

Reply:
The default max retries is set to 1 in new commit, which means the we don't automatically retry but provide retrying methods for users.